src/main/haskell/kore/src/Kore/Step/Pattern.hs

```diff
@@ -327,14 +335,80 @@ may appear in the right-hand side of any substitution, but this is not checked.
 substitute
     ::  ( FreshVariable variable
         , MetaOrObject level
-        , MonadCounter m
         , Ord (variable level)
         , SortedVariable variable
         )
     => Map (variable level) (StepPattern level variable)
     -> StepPattern level variable
-    -> m (StepPattern level variable)
+    -> StepPattern level variable
 substitute = Substitute.substitute (Lens.lens getFreeVariables setFreeVariables)
   where
     getFreeVariables Valid { freeVariables } = freeVariables
     setFreeVariables valid freeVariables = valid { freeVariables }
+
+{- | Reset the 'variableCounter' of all 'Variables'.
+
+@fromFreshVariables@ resets the 'variableCounter' of all variables, while
+ensuring that no 'Variable' in the result is accidentally captured.
+
+ -}
+fromFreshVariables
+    :: forall level. MetaOrObject level
+    => StepPattern level Variable
+    -> StepPattern level Variable
+fromFreshVariables stepPattern =
+    Reader.runReader
+        (Recursive.fold fromFreshVariablesWorker stepPattern)
+        Map.empty
+  where
+    lookupVariable variable =
+        Reader.asks (Map.lookup variable) >>= \case
+            Nothing -> return variable
+            Just variable' -> return variable'
+
+    -- | Refresh a 'Fresh variable level' until it is unique among a set of
+    -- 'Variable level'.
+    safeVariable freeVariables' variable =
+        head
+        $ dropWhile (\var -> Set.member var freeVariables')
+        $ iterate nextVariable variable
+
+    underBinder freeVariables' variable child = do
+        let variable' = safeVariable freeVariables' variable
+        child' <- Reader.local (Map.insert variable variable') child
+        return (variable', child')
+
+    fromFreshVariablesWorker (valid :< patt) = do
```

Comment:
Having type signatures for some of these functions would make the code easier to read.